### ISSUE TITLE
Configure Shortcuts

### ISSUE THREAD

**Issue by Raffinade** [xournalpp/xournalpp#919]
For me it would be fine to configure shortcuts in the frontend even for the colors, so that i dont have to put the pointer on the color to change it.
I hope it will be possible.

Greets Raffinade

**Comment by LittleHuba** [comment]
We should add a section to the settings for this and make all shortcuts we have adjustable.

**Comment by LittleHuba** [comment]
This will require a global system listening for shortcuts. They can not be bound to the buttons themselves anymore (which might be better anyway).

**Comment by relunctance** [comment]
I'm looking forward to this shortcuts/keybindings . When can I used it

**Comment by Narice** [comment]
Yes, that would be a must!

**Comment by ave-63** [comment]
The only reason I am still using MS windows is because it has PDF Annotator, which has keyboard shortcut customization for almost every action. I use lines, four colors of pens, four colors of highlighters, lasso select, copy, paste, undo, delete, and a few other things this way. I have them bound to regular keys (asdfg, ASDF, qwert, etc). 

LittleHuba says they cannot be bound to regular keys in xournalpp, and that is a shame! I'm a math teacher that uses PDF Annotator for regular lectures, and I will get sore fingers from holding down shift+ctrl all day.

Anyway, I was learning how to use git and cmake so that I could implement this myself, but it looks like others are already working on it? It will take me a ton of time to get up to speed because I'm not a software developer.

**Comment by asanakoy** [comment]
My vote for customizable shortctus as well. 
Or at least adding shortcuts for changing Pen line style would be helpful

**Comment by tobiasBora** [comment]
I would also love custom shortcuts, because I'm using several tools (xournalpp, drawpile...) and I'd love to have consistent, short, shorcuts.

But at least, as a first simple thing, would it be possible to bind "Ctrl-Shift-E" to "Export as PDF"? (the shorcut would even be consistent with Gimp shortcuts) I use that feature always and I'm tired of clicking with the mouse.

Thanks a lot!

**Comment by kamdh** [comment]
PgUp and PgDn should automatically go up or down a page. Ctrl-PgUp/PgDn is cumbersome

**Comment by JaviMaciasG** [comment]
I would also like to have all drawing elements to be accessible by keyboard. Most tools are already this way, but I heavily miss this feature for colors and pen options. I am using xournal++ for synchronous online and offline teaching (in a public University) and constant mouse movements to access some drawing features are distracting to students who are used to following my cursor around.

Thank you very much for your superb application!!

**Comment by Febbe** [comment]
I am looking forward to this: https://developer.gnome.org/gtk4/stable/GtkShortcutController.html but it is only gtk4, not gtk3.

**Comment by cuppajoeman** [comment]
I just purchased a drawing tablet, and started binding my buttons to xournal's features, I was making lots of progress until I couldn't find shortcuts for changing colours and pen size. I'm looking forward to this so I can complete my configurations!

**Comment by codethief** [comment]
@Febbe This is probably a dumb question (I've got only limited experience with Gtk) but what's changing compared to how shortcuts how set up in Gtk3?

**Comment by rolandlo** [comment]
I am currently looking at what of the shortcuts requested here can already be done with plugins. A (very recently added) documentation on how to write a plugin can be found on [the new website](https://xournalpp.github.io/guide/plugins/).

- changing line style can easily be achieved with the command   

```lua
app.uiAction({["action"] = "ACTION_TOOL_LINE_STYLE_DASH"})
```
Use LINE_STYLE_DASH_DOT, LINE_STYLE_DOT and LINE_STYLE_PLAIN for the other line styles.
- changing size of the tool in use can be achieved with the command

```lua
app.uiAction({["action"] = "ACTION_SIZE_VERY_FINE"})
```
Use SIZE_FINE, SIZE_MEDIUM, SIZE_THICK and SIZE_VERY_THICK for the other sizes

If you want to specifically change the size for the PEN or HIGHLIGHTER use e.g. 
ACTION_TOOL_PEN_SIZE_VERY_FINE or ACTION_TOOL_HILIGHTER_SIZE_VERY_FINE instead (note the spelling for the HI(GH)LIGHTER, seems we have a typo in the code).

- exporting to pdf (or to another format selectable from the filter on the bottom right) already has the shortcut CTRL+E. 

- (rectangle, lasso and object) select, copy, paste, undo, redo, delete can be achieved by using the actions
ACTION_TOOL_SELECT_RECT, ACTION_TOOL_SELECT_REGION, ACTION_TOOL_SELECT_OBJECT, ACTION_COPY, ACTION_PASTE, ACTION_UNDO, ACTION_REDO, ACTION_DELETE

In general have a loot at [the code](https://github.com/xournalpp/xournalpp/blob/master/src/control/Control.cpp#L371-L947?) for available actions. 

Changing colors seems not yet to be possible via plugins, but I think I will manage to extend the plugin interface to include it.

**Comment by ave-63** [comment]
@rolandlo Wow, the solution was right under our nose this whole time! I went ahead and made [my own plugin](https://github.com/ave-63/xournal-keys) which does most of what I was looking for. I still have some questions, and maybe one bug report. Not sure if this is the right place but here goes:

- I'm not sure what can be done with `app.uiActionSelected`. The documentation is pretty vague, and I could probably figure it out by poring through the C code, but I'm not a programmer and it would be much easier if there were a list of commands. The only way I knew the list of commands for `app.uiAction` is because you linked it here.

- It seems like `ACTION_RULER` has a problem. It seems to work right when I press it once, changing it to ruler mode. But then pressing it a second time leaves it in ruler mode so I have to turn it off with the regular mouse controls.

- `ACTION_RULER`, and a few other commands (`ACTION_TOOL_FILL`, `RECT`, `CIRCLE') seem like they are supposed to function as a toggle. It would be really nice if there were a way to explicitly turn them on or off. I believe the way it is now, I must do `ACTION_RULER` to turn on the ruler, and then remember to do it again when I want a regular pen. It would be nice if I could just press the key for the regular pen (or whatever non-ruler tool I want to use next) and have that turn the ruler off automatically.

Thanks for the work and I look forward to a color-switching interface!

**Comment by rolandlo** [comment]
@ave-63 

- Frankly, I am also not sure, what can be done with `app.uiActionSelected` (what can't be done with `app.uiAction` in a similar or better way). It gets called by `app.uiAction` anyway, it seems. It calls the action listeners which need to be notified about a certain action. In the [toggleGrid plugin](https://github.com/xournalpp/xournalpp/blob/master/plugins/ToggleGrid/togglegrid.lua) the 
```lua 
app.uiActionSelected("GROUP_GRID_SNAPPING", "ACTION_NONE");
``` 
command can be replaced by 

```lua
app.uiAction({["action"] = "ACTION_GRID_SNAPPING", ["enabled"] = false});
```
and similarly if the snapping to grid should be turned on. 

By the way you may pass the "group" key via `["group"] = "GROUP_GRID_SNAPPING"` but as I see from the code, the group key will only used for displaying warnings. It doesn't change the behavior in any other way. So for simplicitly I would just always omit the "group" key.

I haven't written the plugin API myself and I haven't figured out or tested everything yet. Definitively the `app.uiAction` and `app.uiActionSelected` functions need better documentation. It will be improved once I know all the details. Feel free to make a PR to improve the documentation yourself. It would be very welcome.

- The Lua table passed to app.uiAction accepts an "enabled" key. By default it is set to `true`, so you would have to set it to `false` like in

```lua
app.uiAction({["action"] = "ACTION_RULER", ["enabled"] = false})
```
Use a Lua toggle variable to keep track of the current status (c.f. [toggleGrid plugin](https://github.com/xournalpp/xournalpp/blob/master/plugins/ToggleGrid/togglegrid.lua)).

Actually I hadn't realized that one can use several `app.registerUi` functions to define multiple shortcuts in the same plugin. Thanks for figuring that out. This info should get into the documentation as well.

**Comment by rolandlo** [comment]
I have to correct my last post about `app.uiActionSelected`.  
First of all in 
```lua
app.uiAction({["action"] = "ACTION_GRID_SNAPPING", ["enabled"] = false});
```
the `["enabled"] = false` is ignored and can be skipped: 

```lua
app.uiAction({["action"] = "ACTION_GRID_SNAPPING"});
```
The command simply toggles between grid snapping on and off. The reason 

```lua
app.uiActionSelected("GROUP_GRID_SNAPPING", "ACTION_NONE");
```
was used, is that this is supposed to turn/keep grid snapping off regardless of whether it is on or off. 
Unfortunately in contrast to `app.uiAction` it doesn't save the selected (resp. toggled) state in the settings, which apparently makes it non-functional. Or does the toggleGrid plugin (with respect to snapping) work for you? I guess the app.uiActionSelected method should be fixed (or removed). At least I would recommend to do everything you can with the `app.uiAction` function instead of using the `app.uiActionSelected` function.

**Comment by rolandlo** [comment]
I have managed to extend the plugin interface such that colors can be changed. It can be used like:
```lua 
app.changeColor({["color"] = 0xff0077, ["tool"] = "pen"})
```
when you want to change pen color (or color from some other tool with color capabilites like "hilighter", "text"
The 0xff0077 is a hex code for RGB values, it starts with "0x". 

```lua 
app.changeColor({["color"] = 0xff0077, ["selection"] = true})
```
when you want to change the color of the current tool and have the colors of elements from a selection changed if you are currently editing one

A PR will follow soon.

**Comment by rolandlo** [comment]
To everyone here: Having colors changeable once PR #2251 is merged, is there anything else the plugin API should provide, so you can get all the shortcuts you need?

**Comment by rolandlo** [comment]
PR #2251 is merged and colors can be changed via plugin now (in Xournal++ version 1.1.0+dev). The name changed from changeColor to changeToolColor, so you will have to write for instance 
```lua 
app.changeToolColor({["color"] = 0xff0077, ["tool"] = "pen"})
```

**Comment by alkuzman** [comment]
Will this build be available [here](https://copr.fedorainfracloud.org/coprs/luya/xournalpp/) or we should make our own build? Thanks for the update :tada:

**Comment by rolandlo** [comment]
The latest build on that site is 3 months old, so it might take a while until this feature will be available there. You can either build from source or use the latest [nightly release](https://github.com/xournalpp/xournalpp/releases/tag/nightly). This feature is already included there.

**Comment by alkuzman** [comment]
Hi, I made a build from the source but I am getting the following error:
`../../plugins/ColorStyle/main.lua:34: attempt to call a nil value (field 'changeToolColor')` when I call *Cycle through color list* from the Plugin menu or press `<Alt>+c`. I am not sure whether I miss something in the configuration but I can look into the issue deeper later today. I am a newbie on this project and I might need help :)

**Comment by rolandlo** [comment]
That means your plugin API does not have the changeToolColor command. Either you are building from the wrong source (check the luapi_application.h file if the changeToolColor command is listed) or you are executing a version other than the one built from source (you should execute by src/xournalpp from the build folder).

**Comment by alkuzman** [comment]
The first time I tried from build with `./src/xournalpp` but it was missing the Plugin menu. Then I entered the src and I tried `xournalpp` which called the installed version from the fedora repo (Sorry for the confusion and thank you).

However, the problem was that `lua-devel` was missing from the deps for plugins. I followed [this](https://github.com/xournalpp/xournalpp/blob/master/readme/LinuxBuild.md) page but `dnf install lua-devel` is missing for Fedora/CentOS/REHL. 

Works perfectly :+1:

**Comment by rolandlo** [comment]
For those who prefer to have relatively few shortcuts to remember like me, note that you can easily cycle through different tools/colors/line styles/eraser modes etc. using Lua tables and local variables. Take for example this lua file (unzip it to use it):

[main.zip](https://github.com/xournalpp/xournalpp/files/5294687/main.zip)

The plugin provides the following shortcuts:

```lua
  app.registerUi({["menu"] = "Grid snapping on/off", ["callback"] = "gridsnap", ["accelerator"] = "<Alt>g"});
  app.registerUi({["menu"] = "Rotation snapping on/off", ["callback"] = "rotsnap", ["accelerator"] = "<Alt>r"});
  app.registerUi({["menu"] = "Fill on/off", ["callback"] = "fill", ["accelerator"] = "<Alt>f"});
  app.registerUi({["menu"] = "Cycle through tools", ["callback"] = "tool", ["accelerator"] = "<Alt>t"});
  app.registerUi({["menu"] = "Cycle through line style", ["callback"] = "linestyle", ["accelerator"] = "<Alt>s"});
  app.registerUi({["menu"] = "Cycle through eraser type", ["callback"] = "eraser", ["accelerator"] = "<Alt>e"});
  app.registerUi({["menu"] = "Cycle through drawing type", ["callback"] = "drawingtype", ["accelerator"] = "<Alt>v"}); 
  app.registerUi({["menu"] = "Cycle through colors", ["callback"] = "color", ["accelerator"] = "<Alt>c"});
```
which you can easily modify to your needs. Make sure the shortcuts don't interfere with your menu accelerators and system shortcuts. Although I haven't had time to test it much, it does seem to improve the workflow quite a bit.

**Comment by AJJLagerweij** [comment]
I love the shotcut tool, I currently have one issue, the behaviour of "Cycle through drawing types" `<Alt>v` opens for me the View menu. (Its called Weergave in the Dutch translation) instead of cycling the drawing types.
After reassigning it to `<Alt>x` the behaviour was as expeced.
```lua
  app.registerUi({["menu"] = "Cycle through drawing type", ["callback"] = "drawingtype", ["accelerator"] = "<Alt>x"}); 
```
I assume that this has to do with `<Alt>v` already being used in the Dutch translation of Xournal++.

**Comment by rolandlo** [comment]
Yes, the "<Alt> + first letter of menu name" key combination activates the menus from the menubar. Those letters are language specific. The shortcut plugin above works without change for German language, but must be modified for most other languages (like you did for Dutch language).

**Comment by AJJLagerweij** [comment]
Thanks for the confirmation. It's a nice plugin, and editing it for personal preference is nice and simple.

For my convenience I've also included a reverse colorcycler as this allows me to go back to the previous colour quickly.
```lua
app.registerUi({["menu"] = "Cycle back through colors", ["callback"] = "colorback", ["accelerator"] = "<Primary><Alt>c"});

function colorback()
  currentColor = (currentColor - 2) % #colorList + 1
  app.changeToolColor({["color"] = colorList[currentColor][2], ["selection"] = true})
  print("Color: " .. colorList[currentColor][1])
end
```
I mapped the dedicated buttons on my drawing tablet to the shortcuts making the use of xournal++ real nice and quick.

**Comment by julianferres** [comment]
> For those who prefer to have relatively few shortcuts to remember like me, note that you can easily cycle through different tools/colors/line styles/eraser modes etc. using Lua tables and local variables. Take for example this plugin (unzip it and extract it to a subfolder of the plugins folder to use it):
> 
> [ShortcutsPlugin.zip](https://github.com/xournalpp/xournalpp/files/5298436/ShortcutsPlugin.zip)
> 
> The plugin provides the following shortcuts:
> 
> ```lua
>   app.registerUi({["menu"] = "Grid snapping on/off", ["callback"] = "gridsnap", ["accelerator"] = "<Alt>g"});
>   app.registerUi({["menu"] = "Rotation snapping on/off", ["callback"] = "rotsnap", ["accelerator"] = "<Alt>r"});
>   app.registerUi({["menu"] = "Fill on/off", ["callback"] = "fill", ["accelerator"] = "<Alt>f"});
>   app.registerUi({["menu"] = "Cycle through tools", ["callback"] = "tool", ["accelerator"] = "<Alt>t"});
>   app.registerUi({["menu"] = "Cycle through line style", ["callback"] = "linestyle", ["accelerator"] = "<Alt>l"});
>   app.registerUi({["menu"] = "Cycle through selection tools", ["callback"] = "select", ["accelerator"] = "<Alt>s"});
>   app.registerUi({["menu"] = "Cycle through eraser type", ["callback"] = "eraser", ["accelerator"] = "<Alt>e"});
>   app.registerUi({["menu"] = "Cycle through drawing type", ["callback"] = "drawingtype", ["accelerator"] = "<Alt>v"}); 
>   app.registerUi({["menu"] = "Cycle through colors", ["callback"] = "color", ["accelerator"] = "<Alt>c"});
> ```
> 
> which you can easily modify to your needs. Make sure the shortcuts don't interfere with your menu accelerators and system shortcuts. Although I haven't had time to test it much, it does seem to improve the workflow quite a bit.
> 
> **Edit:** simplified the code and added console messages for toggling functions
> **Edit:** added the selection tool and changed the accelerator for line style. Added the plugin.ini file to the zip-file

Hi! I am having problems just with the "Cycle through colors" shorcut, added the message that pop's up when using this particular shorcut

![Screenshot from 2020-11-19 16-43-55](https://user-images.githubusercontent.com/32100099/99716199-95ac1880-2a86-11eb-9578-b0e290a42fe0.png)

**Comment by rolandlo** [comment]
@julianferres That means you don't have version 1.1.0+dev installed. The `changeToolColor` command has only added to the plugin API recently. You can fix that by installing the latest nightly release.

**Comment by julianferres** [comment]
Works like a charm, thank you so much for the instant reply :+1:

**Comment by twitwi** [comment]
I'm cross-referencing:
Currently, binding to a plain key (without modifiers) breaks the text tool as this letter cannot be typed anymore.
One might want to consider the modification of xournalpp suggested in https://github.com/xournalpp/xournalpp/issues/977 to solve this issue (and the one from the issue).

**Comment by rfaridi** [comment]
Hello, I am having the same problem as @julianferres initially had.  Plz check the images for exact problem and my version of xournalpp. I am using arch linux with i3 window manager. 

![image](https://user-images.githubusercontent.com/8208933/105169172-59a83900-5b45-11eb-869a-356031e51984.png)

![image](https://user-images.githubusercontent.com/8208933/105169264-7e9cac00-5b45-11eb-8926-ccf2915e4288.png)

**Comment by rolandlo** [comment]
The answer is still the same. Install version 1.1.0+dev.

**Comment by LittleHuba** [comment]
@rfaridi you will have to use the nightly to use this functionality. Either download it from the release page or through AUR.

**Comment by rfaridi** [comment]
The AUR version seems to be quite old. In that case should I build from the source from Release page? Also should I remove the current stable version first? Sorry for somewhat NOOB  questions.

**Comment by LittleHuba** [comment]
Please remove the current stable version and use this AUR package. It is automatically up to date with our master branch.
https://aur.archlinux.org/packages/xournalpp-git/

**Comment by rfaridi** [comment]
Thanks a lot for  quick response. It works like a charm now.

**Comment by rfaridi** [comment]
Couple of questions:  is there a way to reverse cycle through colors?  Also can I map shortcut keys to single colors? I will only use 3/4 specific colors, so I would like to have specific shortcuts for each color.

**Comment by rfaridi** [comment]
I have seen there is a code in this page for reverse cycle through colors.  Can I just cut and paste that in the plugin lua file?

**Comment by rolandlo** [comment]
Yes, you can cut and paste that code. Make sure you place the `app.registerUi` call into the `initUi` function.
And yes, you can also have specific shortcuts for each color. You have to register every accelerator in the `initUi` function and define the callback function, which then will just contain one call of `app.changeColor`

**Comment by rfaridi** [comment]
Thanks a lot. That was quite helpful.

**Comment by JaviMaciasG** [comment]
Thanks all for the great work in making this work through the plugins. 

I've been using this feature for a while (to change colors and access tools), and now I'm trying to map key presses in the numeric pad to the accelerators in the lua script. Mapping a "3" in the top row of the keyboard works perfectly (GDK_KEY_3 0x033), but pressing "3" in the numeric pad generates a different keyboard event (GDK_KEY_KP_3 0xffb3) and is thus not recognized in the plugin. 

Do you know how can I specify key presses from the numeric pad in the accelerator field of the lua script?

Thank you very much!!

**Comment by rolandlo** [comment]
The accelerator string "KP_3" seems to work for `GDK_KEY_KP_3"

**Comment by JaviMaciasG** [comment]
Works perfectly, thank you very much @rolandlo !!

**Comment by r-menezes** [comment]
I am not being able to configure @rolandlo ShortcutPlugin to include the "default" pen tool while cycling through the possible tools.
I peeked into the source code and I think that with the current version it is not possible, it would be simple to update this, though, just adding a DEFAULT action for the tool in Control.cpp,

something along the lines:

`case ACTION_TOOL_DRAW_DEFAULT:` in L575

`(this->toolHandler->getDrawingType() == DRAWING_TYPE_DEFAULT && type == ACTION_TOOL_DRAW_SPLINE) ||` in L1135

and

```
        case ACTION_TOOL_DRAW_DEFAULT:
            this->toolHandler->setDrawingType(DRAWING_TYPE_DEFAULT);
            break;
```
in L1169-71

If you so wish, I can make a pull request

**Comment by fuzxi** [comment]
@rolandlo 

There's a typo on line 57 of your plugin that causes a crash when changing tools twice. `HILIGHTER` should be `HIGHLIGHTER`.

**Comment by rolandlo** [comment]
@fuzxi 
Thanks for pointing this out. The spelling was corrected in the C++-code from `HILIGHTER` to `HIGHLIGHTER` a couple of months ago. I updated it in the Lua-Plugin now.

**Comment by rolandlo** [comment]
@r-menezes 
The name in the C++ code is `ACTION_TOOL_DEFAULT`, so you should add `"TOOL_DEFAULT"` in `drawingtypeList`

**Comment by Febbe** [comment]
A crash from an external source, is not good and should be guarded. Can you find out, why it is crashing?

**Comment by rolandlo** [comment]
@Febbe It crashes because of the line
https://github.com/xournalpp/xournalpp/blob/1002c131e1d7a1cb1243d955c611979b7faec488/src/enums/generated/ActionType.generated.cpp#L1209
Similarly we have
https://github.com/xournalpp/xournalpp/blob/1002c131e1d7a1cb1243d955c611979b7faec488/src/enums/generated/ActionGroup.generated.cpp#L225

We would have to change the `generateConvert.php` script that produces these lines:
https://github.com/xournalpp/xournalpp/blob/1002c131e1d7a1cb1243d955c611979b7faec488/src/enums/generateConvert.php#L99-L119

**Comment by rolandlo** [comment]
We could replace that `g_error` by `g_warning`, but we will also have to add the case `ACTION_NONE`to the switch statement in`Control::actionPerformed`, otherwise we will run into this:

https://github.com/xournalpp/xournalpp/blob/1002c131e1d7a1cb1243d955c611979b7faec488/src/control/Control.cpp#L936-L939

**Comment by r-menezes** [comment]
Hi @rolandlo,
for me the `TOOL_DEFAULT` technique also changes the width and color of the Pen.
I am using this for now, but it would be nice if changing the tool did not imply in color/width changes. Therefore my suggestion.

But overall, xournal++ is the best tool I could find, and is amazing in so many levels!

**Comment by rolandlo** [comment]
@r-menezes Ok, I get it now. Please make a Pull Request. Generally speaking it sounds like a reasonable suggestion to me.

**Comment by Poohol** [comment]
Buenas tardes rolando

la verdad me interesa mucho su trabajo peeo al momento de adjuntar el pluging no se visualiza en xournall++ de windows
espero su ayuda

**Comment by rolandlo** [comment]
> Buenas tardes rolando
> 
> la verdad me interesa mucho su trabajo peeo al momento de adjuntar el pluging no se visualiza en xournall++ de windows
> espero su ayuda

I will answer in English. 
Into what directory did you copy the plugin? What version of Xournal++ do you use? You need [version 1.1.0+dev](https://github.com/xournalpp/xournalpp/releases/tag/nightly). If you already have versoin 1.1.0+dev you may just need to activate the plugin in the plugin manager. Or is it not listed there? Are any plugins listed there at all?

**Comment by Poohol** [comment]
¿En qué directorio copiaste el complemento? C:\Program Files\Xournal++\ui
¿Qué versión de Xournal ++ usas? xournalpp-1.0.20-windows
no aparece en la lista

**Comment by rolandlo** [comment]
Please install [version 1.1.0+dev](https://github.com/xournalpp/xournalpp/releases/tag/nightly).

**Comment by Poohol** [comment]
ok ya lo puedo visualizar y funciona muy bien salvo el de borrar me gustaría personalizar algunos atajos como lo podria cambiar

**Comment by mcp292** [comment]
Is there a pull request open for this feature?

**Comment by Febbe** [comment]
No, and we won't do any until we migrated to gtk4. The hole keybinding system has changed between gtk3 and gtk4.

**Comment by surprized** [comment]
> No, and we won't do any until we migrated to gtk4. The hole keybinding system has changed between gtk3 and gtk4.

Hope it all goes well with gtk4. Looking forward to this feature too :)

**Comment by Febbe** [comment]
I uploaded the current state of the migration in the gtk4 branch, currently it does not compile. But it did last month, before I began to refactor all Dialogs (currently not working, or freezing the app)

**Comment by All3xJ** [comment]
Any news about this very important feature? It would be very good to have it in Xournalpp! :)

**Comment by quasigod-io** [comment]
I hope this is eventually added, being able to change key binds to allow for quicker switching between common tools would be amazing for helping people be more efficient. I always have one hand on my kb and one holding my stylus, so switching between tools a lot is a bit of a pain.

**Comment by UlyssesZh** [comment]
Why should the shortcut for pen be `Ctrl+Shift+P`, where `P` is a key pressed by the right hand? People are usually right-handed, so the right hand has its task to take the physical pen but not to hit the `P` key.

**Comment by mjg** [comment]
> Why should the shortcut for pen be `Ctrl+Shift+P`, where `P` is a key pressed by the right hand? People are usually right-handed, so the right hand has its task to take the physical pen but not to hit the `P` key.

It is `P` as in **P**en. After all, there is no point in shortcuts which you can't remember. Admittedly, there is no point in shortcuts which you can't press either ;-)

In the end, the choice is personal, and so we need an easy way to configure things. Currently, the only way is via plugins in a **system folder**. Also, we probably want to configure key bindings without having to write a plugin. Changing all of that requires quite a few changes and a good design decision.

**Comment by Febbe** [comment]
> > Why should the shortcut for pen be `Ctrl+Shift+P`, where `P` is a key pressed by the right hand? People are usually right-handed, so the right hand has its task to take the physical pen but not to hit the `P` key.
> 
> It is `P` as in **P**en. After all, there is no point in shortcuts which you can't remember. Admittedly, there is no point in shortcuts which you can't press either ;-)
> 
> In the end, the choice is personal, and so we need an easy way to configure things. Currently, the only way is via plugins in a **system folder**. Also, we probably want to configure key bindings without having to write a plugin. Changing all of that requires quite a few changes and a good design decision.

Exactly. There is also the problem, that gtk4 will change a lot regarding the event handling and the shortcut management. Therefore every line of code is a waste of time and effort.

**Comment by Febbe** [comment]
> > Why should the shortcut for pen be `Ctrl+Shift+P`, where `P` is a key pressed by the right hand? People are usually right-handed, so the right hand has its task to take the physical pen but not to hit the `P` key.
> 
> It is `P` as in **P**en. After all, there is no point in shortcuts which you can't remember. Admittedly, there is no point in shortcuts which you can't press either ;-)
> 
> In the end, the choice is personal, and so we need an easy way to configure things. Currently, the only way is via plugins in a **system folder**. Also, we probably want to configure key bindings without having to write a plugin. Changing all of that requires quite a few changes and a good design decision.

Exactly. There is also the problem, that gtk4 will change a lot regarding the event handling and the shortcut management. Therefore every line of code is a waste of time and effort.

**Comment by rolandlo** [comment]
@vladQll 
The plugin works for me (on Ubuntu 22.04). I turn on the eraser and then I can switch between standard eraser and delete stroke eraser via the shortcut `Alt+i`. Note that if the active tool is not the eraser tool, the plugin won't switch to the eraser tool, it just changes the mode in which the eraser tool (once you use it) operates. You could add 
```lua
app.uiAction({["action"] = "ACTION_TOOL_ERASER"})
```
to the `eraser()` callback function, though, if you like to switch the tool automatically. Otherwise please describe what you expect the plugin to do exactly.
By the way what version of Xournal++ do you have installed?

**Comment by rolandlo** [comment]
I've tried out the usage of [Fly-Pie](https://github.com/Schneegans/Fly-Pie) through a plugin. I started with [this plugin](https://github.com/baf57/Xournalpp_FlyPie) by @baf57 and made some corrections and adjustments. Now it works like that:

https://user-images.githubusercontent.com/40485433/183128299-464d67da-7dfa-48fc-8544-f114b2d5e332.mp4

Is anyone interested that this is developed further?

**Comment by vladQll** [comment]
> @vladQll El complemento me funciona (en Ubuntu 22.04). Enciendo el borrador y luego puedo cambiar entre el borrador estándar y borrar el borrador de trazo a través del atajo `Alt+i`. Tenga en cuenta que si la herramienta activa no es la herramienta de borrado, el complemento no cambiará a la herramienta de borrado, solo cambia el modo en que opera la herramienta de borrado (una vez que la usa). podrías agregar
> 
> ```lua
> app.uiAction({["action"] = "ACTION_TOOL_ERASER"})
> ```
> 
> Sin embargo, a la `eraser()`función de devolución de llamada si desea cambiar la herramienta automáticamente. De lo contrario, describa qué espera que haga exactamente el complemento. Por cierto, ¿qué versión de Xournal++ tienes instalada?

I have installed Version 1.1.1+dev, **describe what exactly you expect the add-on to do?**: for example with the selection function, if I press the assigned keys they change the type of selection and I can use them, however that does not happen with the keys assigned to the delete function, not even the eraser icon that would be indicating that I am using it appears.

**Comment by vladQll** [comment]
> ```lua
> app.uiAction({["action"] = "ACTION_TOOL_ERASER"})
> ```

that tool would be great

**Comment by vladQll** [comment]
Hi I have a question @rolandlo when I use the `drawingtype()` function I usually have it assigned to `<Alt>d` to draw arrows or circles quickly and then I use a button linked to `<Alt>k` which calls `tool()` to cycle to the pencil . When I get to the pencil tool it is still selected with the above features like drawing lines or if I previously used the lineStyle cycle. I would like to know what I should add to the `tool()` function so that when I cycle through and get to the pencil tool I can use normal pencil.

**Comment by rolandlo** [comment]
> > ```lua
> > app.uiAction({["action"] = "ACTION_TOOL_ERASER"})
> > ```
> 
> that tool would be great , can be added to windows10 or only works for linux?

That works on Windows 10, too.

**Comment by rolandlo** [comment]
> Hi I have a question @rolandlo when I use the `drawingtype()` function I usually have it assigned to `<Alt>d` to draw arrows or circles quickly and then I use a button linked to `<Alt>k` which calls `tool()` to cycle to the pencil . When I get to the pencil tool it is still selected with the above features like drawing lines or if I previously used the lineStyle cycle. I would like to know what I should add to the `tool()` function so that when I cycle through and get to the pencil tool I can use normal pencil.

You could add something like
```lua
  if (toolList[currentTool] == "PEN") then
    app.uiAction({["action"] = "ACTION_SHAPE_RECOGNIZER"})
    app.uiAction({["action"] = "ACTION_TOOL_LINE_STYLE_PLAIN"})
  end
``` 
to the `tool()` callback function.

**Comment by vladQll** [comment]
> 

me ah funcionado gracias @rolandlo  aunque no se si lo puse en el lugar correcto dentro de la funcion `tool()`

**Comment by vladQll** [comment]
> Probé el uso de [Fly-Pie](https://github.com/Schneegans/Fly-Pie) a través de un complemento. Empecé con [este complemento](https://github.com/baf57/Xournalpp_FlyPie) por@baf57e hizo algunas correcciones y ajustes. Ahora funciona así:
> 
>  DemoFlyPie.mp4 
> ¿Alguien está interesado en que esto se desarrolle más?
I was referring to this tool if it is available for windows I will have to try it.

**Comment by rolandlo** [comment]
> I was referring to this tool if it is available for windows I will have to try it.

Unfortunately Fly-Pie is only available on Gnome-based Linux distributions. So that won't work. There are alternatives on Windows, but none of those I have seen look as nice and configurable as Fly-Pie.

**Comment by vladQll** [comment]
> > 
> 
> me ah funciono gracias@rolandlo aunque no se si lo puse en el lugar correcto dentro de la funcion`tool()` ![imagen](https://user-images.githubusercontent.com/28522011/183157671-9f99dd7a-2f74-4f61-b415-8a609ce36678.png)

a query and is there a way to do it just by calling the normal pencil without calling the shape recognizer , I tried to remove the line `app.uiAction({["action"] = "ACTION_SHAPE_RECOGNIZER"})` but it does not work as I would like it happens the following only works when I have previously used other line styles and then scroll to the pencil tool and this works great great , but if I have previously used tools like arrow , straight lines, etc.. and I go back and choose the pencil this still continues with the previous characteristics or properties (arrows, lines, etc.), then you may wonder but the shape recognizer solves that, actually apparently it does visually but as it is activated when I start writing begins to recognize the shapes of letters such as o, L

**Comment by rolandlo** [comment]
> a query and is there a way to do it just by calling the normal pencil without calling the shape recognizer ,

Depending how your default tool is configured in the preferences, you could try replacing `ACTION_SHAPE_RECOGNIZER` by `ACTION_TOOL_DEFAULT`.

**Comment by vladQll** [comment]
> > una consulta y hay alguna manera de hacerlo simplemente llamando al lápiz normal sin llamar al reconocedor de formas,
> 
> Dependiendo de cómo esté configurada su herramienta predeterminada en las preferencias, puede intentar reemplazarla `ACTION_SHAPE_RECOGNIZER`por `ACTION_TOOL_DEFAULT`.

great that if that worked, of course you are right you previously set the default tool now if it is working thank you very much

**Comment by vladQll** [comment]
@rolandlo hola como esta , tengo una duda hace unos dias agregue estas lineas de codigo para tres colores que uso mucho 
![image](https://user-images.githubusercontent.com/28522011/183643188-001f63fe-9fb4-4413-8e53-b0b5f4bf9313.png)
funciona bien el detalle es que luego de un uso prolongado de alt + 1 , 2 o 3 es un poco molesto , en su lugar de alt quisiera que fuera la tecla Tab para asignarlo y usar tab+1 , tab+2 o tab+3 para negro , azul y rojo respectivamente , pero no se como deberia ser la declaracion mi intuicion dice que quizas es  `"<tab>1"` `"<tab>2"` `"<tab>3"` pero no se si es correcto eso

**Comment by twitwi** [comment]
Thanks for making some of us discover Fly-Pie.
On my side I couldn't find how (if possible) to install plugins in a snap version of xournal++, and I ended up finding that there are experimental features around having a quick tool change https://github.com/xournalpp/xournalpp#experimental-features (which currently is not very convenient).

**Comment by rolandlo** [comment]
@twitwi 
Once we release version 1.2 you can put the plugins in the folder `~/snap/xournalpp/current/.config/xournalpp/plugins` folder, I think. In version 1.1.1 that is not possible with the snap version, unfortunately.

**Comment by rolandlo** [comment]
> @rolandlo hello how are you, I have a question a few days ago I added these lines of code for three colors that I use a lot ![image](https://user-images.githubusercontent.com/28522011/183643188-001f63fe-9fb4-4413-8e53-b0b5f4bf9313.png) works fine the detail is that after prolonged use of alt + 1 , 2 or 3 is a bit annoying , instead of alt I would like to be the Tab key to assign it and use tab+1 , tab+2 or tab+3 for black , blue and red respectively , but I do not know how should be the code my intuition says that maybe is `"<tab>1"` `"<tab>2"` `"<tab>3"` but I do not know if that is correct that

In contrast to `Alt`, `tab` is not a modifier key, so you can't use `tab+1`.

**Comment by twitwi** [comment]
> @twitwi Once we release version 1.2 you can put the plugins in the folder `~/snap/xournalpp/current/.config/xournalpp/plugins` folder, I think. In version 1.1.1 that is not possible with the snap version, unfortunately.

Great news. No problem, I'll install a ppa version and should get the plugin running. Thanks.

I don't want to open an issue (but I can if it makes sense) but would there be a dbus interface (to avoid going through emitting custom shortcuts that are handled by some plugins)?

**Comment by rolandlo** [comment]
> I don't want to open an issue (but I can if it makes sense) but would there be a dbus interface (to avoid going through emitting custom shortcuts that are handled by some plugins)?

Please open an issue about it and give some details. That's nothing I have thought of, but it sounds interesting.

**Comment by vishaltanwar96** [comment]
> I've tried out the usage of [Fly-Pie](https://github.com/Schneegans/Fly-Pie) through a plugin. I started with [this plugin](https://github.com/baf57/Xournalpp_FlyPie) by @baf57 and made some corrections and adjustments. Now it works like that:
> 
>  DemoFlyPie.mp4 
> Is anyone interested that this is developed further?

Hello, thanks for the suggestion. Could you please let us know what are all the adjustments and corrections you've made? I am having a bit of trouble in figuring out how to get the color selection and a lot of other stuff working.

**Comment by dov** [comment]
For what it is worth, here is another gtk program (though written in Vala) with customizable keybindings: https://pdfpc.github.io/ . How difficult would it be to copy their infrastructure?

**Comment by web10027** [comment]
> For those who prefer to have relatively few shortcuts to remember like me, note that you can easily cycle through different tools/colors/line styles/eraser modes etc. using Lua tables and local variables. Take for example this plugin (unzip it and extract it to a subfolder of the plugins folder to use it):
> 
> [ShortcutsPlugin.zip](https://github.com/xournalpp/xournalpp/files/6330702/ShortcutsPlugin.zip)
> 
> The plugin provides the following shortcuts:
> 
> ```lua
>   app.registerUi({["menu"] = "Grid snapping on/off", ["callback"] = "gridsnap", ["accelerator"] = "<Alt>g"});
>   app.registerUi({["menu"] = "Rotation snapping on/off", ["callback"] = "rotsnap", ["accelerator"] = "<Alt>r"});
>   app.registerUi({["menu"] = "Fill on/off", ["callback"] = "fill", ["accelerator"] = "<Alt>f"});
>   app.registerUi({["menu"] = "Cycle through tools", ["callback"] = "tool", ["accelerator"] = "<Alt>t"});
>   app.registerUi({["menu"] = "Cycle through line style", ["callback"] = "linestyle", ["accelerator"] = "<Alt>l"});
>   app.registerUi({["menu"] = "Cycle through selection tools", ["callback"] = "select", ["accelerator"] = "<Alt>s"});
>   app.registerUi({["menu"] = "Cycle through eraser type", ["callback"] = "eraser", ["accelerator"] = "<Alt>e"});
>   app.registerUi({["menu"] = "Cycle through drawing type", ["callback"] = "drawingtype", ["accelerator"] = "<Alt>v"}); 
>   app.registerUi({["menu"] = "Cycle through colors", ["callback"] = "color", ["accelerator"] = "<Alt>c"});
> ```
> 
> which you can easily modify to your needs. Make sure the shortcuts don't interfere with your menu accelerators and system shortcuts. Although I haven't had time to test it much, it does seem to improve the workflow quite a bit.
> 
> **Edit:** simplified the code and added console messages for toggling functions **Edit:** added the selection tool and changed the accelerator for line style. Added the plugin.ini file to the zip-file **Edit:** updated the spelling of `HIGHLIGHTER`

First-time caller, long-time fan of xournal++ .  

I like this plugin.
It has been straightforward to modify it, but there is one issue I cannot figure out.
How can I select a straight line, rather than a box, arrow, or ellipse, in drawingTypeList?
"TOOL_DRAW_LINE" does not work for me and I did not find something like "ACTION_TOOL_DRAW_LINE" in the source code, here.

https://github.com/xournalpp/xournalpp/blob/1002c131e1d7a1cb1243d955c611979b7faec488/src/enums/generated/ActionType.generated.cpp

Thanks !

**Comment by rolandlo** [comment]
@web10027 What you are looking for is called `ACTION_RULER`, I think. It has an inconsistent name, we should probably change that.

**Comment by web10027** [comment]
Now it works !  Thank you, @rolandlo

**Comment by vladQll** [comment]
> @vishaltanwar96En lugar de usar el archivo `src/menu.json`de [ https://github.com/baf57/Xournalpp_FlyPie ](https://github.com/baf57/Xournalpp_FlyPie) ) comience con el `menu.json`archivo en este archivo zip: [ menu.zip ](https://github.com/xournalpp/xournalpp/files/9496911/menu.zip) Ahora use un editor de texto (o el `sed`comando) para reemplazar todas las apariciones de
> 
>     * `XPP_PATH`con el camino a tu `xournalpp`instalación 
>       (normalmente `/usr/share/xournalpp`)
> 
>     * `FP_PATH`con la ruta a la que clonó el `Xournalpp_FlyPie`repositorio 
>       (p.ej `/home/USERNAME/Documents/Xournalpp_FlyPie`)
> 
> 
> Luego proceda como se describe en el [ Léame ](https://github.com/baf57/Xournalpp_FlyPie/blob/main/README.md) . Después de eso, las cosas deberían funcionar.

hola intente todos los pasos pero no logro hacerlo funcionar , estoy en fedora36 , y como queria instalar la ultima version nightly de xournal++ , tuve que descargar el AppImage ya que no hay una forma de instalarla la version en fedora excepto la estable , en resumen , segui los pasos mencionados pero no logro hacer lo que hace en el video , cree que se deba a que estoy usando un appImage?

**Comment by JohnnyLAmpAz** [comment]
This lack of functionality was bothering me so much and I love using Xournal++ for my notes. I suggest you to check out [this plugin](https://github.com/raw-bacon/vi-xournalpp) which might satisfy your needs, like it did for me.

**Comment by PraxTube** [comment]
How could you change the zoom with the shortcut plugin (how can you call it, like with `app.uiAction` or from where)?

**Comment by baf57** [comment]
@PraxTube 
> How could you change the zoom with the shortcut plugin (how can you call it, like with `app.uiAction` or from where)?

`app.uiAction({["action"]="ACTION_ZOOM_IN"})`. If you want info on all the action strings you can pass to the `app.uiAction` method, see the cases in [the `Control::actionPerformed` method](https://github.com/xournalpp/xournalpp/blob/c07654780933929a92e9187ad0dc44a80fb04cc7/src/core/control/Control.cpp#L360-L950).

For anyone generally looking for ways to make shortcuts and plugins, see [the plugin api manual](https://xournalpp.github.io/guide/plugins/plugins/#plugin-api) for more details.

**Comment by PraxTube** [comment]
Thanks, that's exactly what I was looking for.

**Comment by LOOPEER111** [comment]
> We should add a section to the settings for this and make all shortcuts we have adjustable.

At Least, make it more simple to  press. I need press Ctrl+Shift+P whith 3 fingers to change to pen. may P-pen, E-erase, H-Highlight is better.

**Comment by onlycparra** [comment]
Inspired by @web10027 , I created my own little plugin that uses `Alt+` {`1`, `2`, `3`, `4`} to cover most of the tools. And `Alt+` {`a`, `o`, `q`} for grid/rotation snapping, and fill (those keys make sense in a dvorak keyboard, feel free to change them). Colors are already accessible through simple numbers from 1 to 0, so I just focused in the tools:

- `Alt+1` cycles through: pen, eraser, and highlighter.
- `Alt+2` cycles through: select rectangle, and select region.
- `Alt+3` selects pen and cycles through: handwriting, straight line, rectangle, arrow, ellipse.
- `Alt+4` selects pen and cycles through: solid, dashed, and dotted line.
- `Alt+a` toggles grid snapping.
- `Alt+o` toggles rotation snapping.
- `Alt+q` toggles shape filling.

If you are currently in (for example) 'select region' and type `Alt+1`, it goes back to the last used drawing tool (pen, eraser, or highlighter). Typing `Alt+1` again will pick up the cycling from it. This feature allows you to use `Alt+1` and `Alt+2` to jump back and forth between your favorite drawing tool and selection tool.

As you may know, to install it (on linux) create the folder `/usr/share/xournalpp/plugins/ToolCycle/` and copy the `main.lua` and `plugin.ini` files in there.

[ToolCycle.zip](https://github.com/user-attachments/files/16775789/ToolCycle.zip)

**Comment by BlindAndInsane** [comment]
any updates regarding this. This the only feature that i miss very much.

**Comment by onlycparra** [comment]
@BlindAndInsane did you try my plugin? you can open the lua file and change the keys if you want. They have to be <Alt> + something, though

**Comment by BlindAndInsane** [comment]
@onlycparra  your plugin is cool but thats not what i am looking for. i want to change key bind for my pen and eraser to be p and e and other tools like that. like i use two apps and i dont wanna change my button on my pen everytime i switch and sadly both dont have keybind settings. i wish xournallpp gets this feature it will just make it sooo good for me. i would have contributed to the project but i dont think my C++ skills are good enuf yet.

**Comment by dov** [comment]
> @onlycparra your plugin is cool but thats not what i am looking for. i want to change key bind for my pen and eraser to be p and e and other tools like that. like i use two apps and i dont wanna change my button on my pen everytime i switch and sadly both dont have keybind settings. i wish xournallpp gets this feature it will just make it sooo good for me. i would have contributed to the project but i dont think my C++ skills are good enuf yet.

I agree! I'm using Krita a lot, and would love to have the same bindings in xournalpp as in Krita. In my case B for brush (pen) and E for the eraser.

**Comment by Mr-FuzzyPenguin** [comment]
I made a pretty rudimentary plugin and quickly-put-together script that can be found in here: https://github.com/Mr-FuzzyPenguin/Pen-GUI-n

Clone it in `/usr/share/xournalpp/plugins/` to get the plugin on your system (Linux).

**Comment by theCarefulIcarus** [comment]
Hey there!
I'm thinking about which plugin to install (for windows), as there have been several options posted. Is there a recommended solution to modify keyboard shortcuts?
Also I'm mainly using the textbox to add notes and the underlining or highlighting of marked text. Are there any (existing) keyboard shortcuts for highlighting text in the pdf? Or are there methods in the API to create shortcuts for said 2 functions?

I couldn't find them in the code and currently the only way I know how to achieve what I intend is to select the text with ctrl+shift+w and then navigate the tooltip menu by mouse. Having a designated tool that highlights printed text would be really userfriendly and cool. Same for underlining. Is there sth I'm missing or potential feature additions?
Thanks to the devs in advance!

**Comment by Mr-FuzzyPenguin** [comment]
Hi! I'm not a developer of Xournal++ but thanks for the suggestion for Pen-GUI-n. I will look at it in greater detail and see if it is possible.

**Comment by SidharthaDR** [comment]
@Mr-FuzzyPenguin Hello First off, thank you for the pen-gui-n plugin for Xournalpp!

---
**Subject:** "Idea for new feature"
**category:** "Visual guide" 
**How it helps:** "**Enables user** to know in **which mode they are in** and **what key they pressed** and **what it does.**"

**Body:**
Can you add a **feature** for your **same plugin**: **pen-gui-n**, which **shows [[mode], [shortcut pressed], [short description of pressed shortcut]]** as **toast for few seconds then dissappear.**

---

**Examples:**
eg1- when i **click on "p"** it shows toast message of [**mode**: page, **shortcut**: default, **Descrp:** "In Page mode"] for few seconds.
eg2- When i **click on "a"** it shows toast message like [**mode**:page, **shortcut**: a, **Descrp:** "new page"] for few seconds.

---
**Note:**
It **doesn't have to follow this [**mode**: " ", **shortcut**:" " , **Descrp:** " "] syntax** you **can use your own idea**. Goal is **user needs to understand** "in which **mode they are in and short description of how the key they pressed helps or works"**

**Comment by atticus-sullivan** [comment]
@SidharthaDR Note: this request is closely related to (if no duplicate of) https://github.com/xournalpp/xournalpp/issues/6017 (except it's no toast notifiers)

**Comment by iFabse** [comment]
Do I have to use the dev's version in order to use the FlyPie plugin, because my selections doesn't do anything to my xournal. Like switching colors ...

**Comment by iFabse** [comment]
After I used the forked project from @Settordici, FlyPie is up and running, I then configured my Lenovo Digital Pen 2 Stylus with input remapper, so I can press one stylus putton and the FlyPie opens

**Comment by UltraBlackLinux** [comment]
any updates on this? I hate having to move my mouse all the time. It's incredibly annoying

**Comment by Hicortab** [comment]
@rolandlo It's already resolved, could you close it?
